lib/Optimizer/Dialect/FIROps.cpp

```diff
@@ -0,0 +1,573 @@
+                         llvm::SmallVectorImpl<llvm::StringRef> &visited) {
+  if (auto st = inType.dyn_cast<fir::SequenceType>()) {
+    auto shape = st.getShape();
+    if (shape.size() == 0)
+      return true;
+    for (auto ext : shape)
+      if (ext < 0)
+        return true;
+  } else if (auto rt = inType.dyn_cast<fir::RecordType>()) {
+    // don't recurse if we're already visiting this one
+    for (auto name : visited)
+      if (name == rt.getName())
+        return false;
+    // keep track of record types currently being visited
+    visited.push_back(rt.getName());
+    for (auto &field : rt.getTypeList())
+      if (verifyInType(field.second, visited))
+        return true;
+    visited.pop_back();
+  } else if (auto rt = inType.dyn_cast<fir::PointerType>()) {
+    return verifyInType(rt.getEleTy(), visited);
+  }
+  return false;
+}
+
+static bool verifyRecordLenParams(mlir::Type inType, unsigned numLenParams) {
+  if (numLenParams > 0) {
+    if (auto rt = inType.dyn_cast<fir::RecordType>())
+      return numLenParams != rt.getNumLenParams();
+    return true;
+  }
+  return false;
+}
+
+// AllocaOp
+
+mlir::Type AllocaOp::getAllocatedType() {
+  return getType().cast<ReferenceType>().getEleTy();
+}
+
+/// Create a legal memory reference as return type
+mlir::Type AllocaOp::wrapResultType(mlir::Type intype) {
+  // FIR semantics: memory references to memory references are disallowed
+  if (intype.dyn_cast<ReferenceType>())
+    return {};
+  return ReferenceType::get(intype);
+}
+
+mlir::Type AllocaOp::getRefTy(mlir::Type ty) { return ReferenceType::get(ty); }
+
+// AllocMemOp
+
+mlir::Type AllocMemOp::getAllocatedType() {
+  return getType().cast<HeapType>().getEleTy();
+}
+
+mlir::Type AllocMemOp::getRefTy(mlir::Type ty) { return HeapType::get(ty); }
+
+/// Create a legal heap reference as return type
+mlir::Type AllocMemOp::wrapResultType(mlir::Type intype) {
+  // Fortran semantics: C852 an entity cannot be both ALLOCATABLE and POINTER
+  // 8.5.3 note 1 prohibits ALLOCATABLE procedures as well
+  // FIR semantics: one may not allocate a memory reference value
+  if (intype.dyn_cast<ReferenceType>() || intype.dyn_cast<HeapType>() ||
+      intype.dyn_cast<PointerType>() || intype.dyn_cast<FunctionType>())
+    return {};
+  return HeapType::get(intype);
+}
+
+// BoxDimsOp
+
+/// Get the result types packed in a tuple tuple
+mlir::Type BoxDimsOp::getTupleType() {
+  llvm::SmallVector<mlir::Type, 3> triple{
+      getResult(0).getType(), getResult(1).getType(), getResult(2).getType()};
+  return mlir::TupleType::get(triple, getContext());
+}
+
+// CallOp
+
+void printCallOp(mlir::OpAsmPrinter &p, fir::CallOp &op) {
+  auto callee = op.callee();
+  bool isDirect = callee.hasValue();
+  p << op.getOperationName() << ' ';
+  if (isDirect)
+    p << callee.getValue();
+  else
+    p << op.getOperand(0);
+  p << '(';
+  p.printOperands(llvm::drop_begin(op.getOperands(), isDirect ? 0 : 1));
```

Comment:
You can do `p << ... << op.getOperands().drop_begin(...)`

Reply:
Tried this, but the compiler asks if I meant `drop_back` instead of `drop_begin`.

Reply:
Ahh, sorry I forgot that the naming is different: drop_front:
https://github.com/llvm/llvm-project/blob/0b6abe428164c38c7dec6c539e454cfd671067a4/mlir/include/mlir/Support/STLExtras.h#L255

Most ranges in MLIR use the above as a base, so most of the ArrayRef functionality is available automatically.